docs/DevGuide/BuildSystem.md

```diff
@@ -4,39 +4,98 @@ See LICENSE.txt for details.
+
+SpECTRE organizes source files into subdirectories of `src` that are
+compiled into libraries.  To add a new source file `FILE.cpp` to an
+existing library in `src/PATH/DIR`, just edit
+`src/PATH/DIR/CMakeLists.txt` and add `FILE.cpp` to the list of files
+in
+```
+set(LIBRARY_SOURCES
+  <list_of_files>
+  )
+```
+such that the resulting `<list_of_files>` is in alphabetical order.
+
+### Adding Libraries {#adding_libraries}
+
+To add a source file `FILE.cpp` that is compiled into a new library `LIB` in a
+directory `src/PATH/DIR` (either in a new directory, or in an existing
+directory that either does not have a `CMakeLists.txt` file, or does
+not create a library in the existing `CMakeLists.txt`):
+- Create (if necessary) a `CMakeLists.txt` file in `DIR`, with the following
+two lines at the top:
+```
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+```
+- In the parent directory (i.e. `src/PATH`), (if necessary) add the
+following line to its `CMakeLists.txt` file (if necessary, recursively
+do the previous step and this one until you reach a `CMakeLists.txt` that
+adds the appropriate subdirectory):
+```
+add_subdirectory(DIR)
+```
+If there are already other `add_subdirectory()` lines in the file, place
+the new one so that the subdirectories are in alphabetical order.
+- Add the line:
+```
+set(LIBRARY LIB)
+```
+where convention is that `LIB` = `DIR`.  As library names must be
+unique, this is not always possible, in which case the convention is to
+prepend the parent directory to `DIR`.
+- Add the lines
+```
+set(LIBRARY_SOURCES
+  FILE.cpp
+  )
+
+add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
+
+target_link_libraries(
+  ${LIBRARY}
+  <list_of_interfaces>
+  )
+```
+where `<list_of_interfaces>` is an alphabetized list of lines of the form
+```
+  INTERFACE SOME_LIBRARY
+```
+where `SOME_LIBRARY` is a library that must be linked in order for the
+new library to be successfully linked in an executable.
+
+## Adding Unit Tests {#adding_unit_tests}
+
+We use the [Catch](https://github.com/philsquared/Catch) testing
+framework for unit tests. All unit tests are housed in `tests/Unit`
+with subdirectories for each subdirectory of `src`. Add the `cpp` file
+to the appropriate subdirectory and also to the `CMakeLists.txt` in
+that subdirectory. Inside the source file you can create a new test by
+adding a `SPECTRE_TEST_CASE("Unit.Dir.Component",
+"[Unit][Dir][Tag]")`. The `[Tag]` is optional and you can have more
+than one, but the tags should be used quite sparingly.  The purpose of
+the tags is to be able to run all unit tests or all tests of a
+particular set of components, e.g. `ctest -L Data` to run all tests
+inside the `Data` directory. Please see \ref writing_unit_tests
+"writing unit tests", other unit tests and the [Catch
+documentation](https://github.com/philsquared/Catch) for more help on
+writing tests. Unit tests should take as short a time as possible,
+with a goal of less than two seconds.
 
```

Comment:
Should we mention here that we also want to limit the number of distinct test cases because of the Catch initialization time on certain systems?

Reply:
I think it's Charm++'s overhead, not catch (I haven't studied this in detail but the overhead is larger on supercomputers and the only thing that changes in Charm++), but I agree :)